### PR TITLE
Define `returncode_t`, add helper functions

### PR DESCRIPTION
**Pull request by bradjc**
This PR adds `returncode_t` and makes it the standard for all of libtock-c. The expectation is that all functions that can return an error always and only return a `returncode_t`. Any values are returned through pointers passed to the function. Any syscall return values should be converted to a `returncode_t` as soon as possible.

Summary of changes:
- Use helper functions rather than have different drivers manually handle `syscall_return_t` types.
- Return only return codes from functions like `led_count()`.
- No longer use `TOCK_SUCCESS`, instead make it `RETURNCODE_SUCCESS` to note that we are using the return code numbers, _not_ the error code values.
- A little varied cleanup to make the code generally more consistent.

### CODE REVIEW

File: examples/courses/2018-11-SenSys/sensys_udp_rx/main.c [hunk 1]
```diff
@@ -84,16 +84,16 @@ int main(void) {
...
-    case TOCK_EINVAL:
+    case RETURNCODE_EINVAL:
       printf("The address requested is not a local interface\n");
       break;
-    case TOCK_EBUSY:
+    case RETURNCODE_EBUSY:
       printf("Another userland app has already bound to this addr/port\n");
       break;
-    case TOCK_ERESERVE:
+    case RETURNCODE_ERESERVE:
```
Comment (hudson-ayers) [inline examples/courses/2018-11-SenSys/sensys_udp_rx/main.c:96]: For the courses, do we want to leave these as-they-were at the time of the course, or update them?

File: examples/l3gd20/main.c [hunk 2]
```diff
@@ -13,25 +13,29 @@ static int decimals (float f, int n)
...
+
+  if (!l3gd20_is_present ()) {
+    printf ("L3GD20 sensor driver not present\n");
+    return -1;
+  }
+
+  if (l3gd20_power_on ()) {
+    printf ("L3GD20 device is not present\n");
+    return -1;
+  }
```
Comment (hudson-ayers) [inline examples/l3gd20/main.c:25]:
```suggestion
  if (!l3gd20_power_on ()) {
    printf ("L3GD20 device is not present\n");
    return -1;
  }
```

File: libtock/tock.c [hunk 3]
```diff
@@ -35,22 +35,102 @@ int tock_enqueue(subscribe_upcall cb, int arg0, int arg1, int arg2, void* ud) {
...
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U64_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U32) {
+    return RETURNCODE_FAIL;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U32_U32) {
+    return RETURNCODE_FAIL;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U64) {
+    return RETURNCODE_FAIL;
+  }
```
Comment (hudson-ayers) [inline libtock/tock.c:68]: I think all these cases should probably just be replaced by `return RETURNCODE_EBADRVAL` -- if someone calls this function with the wrong comand_return type, it shouldn't "kinda work"

File: libtock/tock.c [hunk 4]
```diff
@@ -35,22 +35,102 @@ int tock_enqueue(subscribe_upcall cb, int arg0, int arg1, int arg2, void* ud) {
...
```
Comment (hudson-ayers) [inline libtock/tock.c:100]: same here

File: libtock/tock.c [hunk 5]
```diff
@@ -35,22 +35,102 @@ int tock_enqueue(subscribe_upcall cb, int arg0, int arg1, int arg2, void* ud) {
...
+  // If the subscribe was successful, easily return SUCCESS.
+  if (subscribe_return.success) {
+    return RETURNCODE_SUCCESS;
+  } else {
+    // Not success, so return the proper returncode.
+    return tock_status_to_returncode(subscribe_return.status);
+  }
+}
+
+int tock_allow_rw_return_to_returncode(allow_rw_return_t allow_return) {
```
Comment (hudson-ayers) [inline libtock/tock.c:117]: in general, I think all these helper functions are fantastic and something we should have done from the start.

**Review by hudson-ayers — CHANGES_REQUESTED** [review]
I have looked at 24/137 files (though I did jump straight to tock.c and tock.h)

Overall I think this looks very good. It should resolve any concerns with the `ReturnCode` --> `ErrorCode` transition, and the helper functions for handling system call results are a very nice convenience.

I did find a few issues so far

File: examples/courses/2018-11-SenSys/sensys_udp_rx/main.c (hunk 1 above)
Comment (bradjc) [reply examples/courses/2018-11-SenSys/sensys_udp_rx/main.c:96]: I think we have to update them for build_all.sh to pass. Also, over time it seems like there is little value (other than for posterity, which git should already handle) to maintain guides for outdated versions of Tock. I'm all for moving to a model where all of our tutorials are always updated as tock is developed.

**Comment by bradjc** [comment]
I agree with the BADRVAL cases and updated the functions.

**Comment by hudson-ayers** [comment]
I propose the following review split. I split it up alphabetically, but with a seperate split for the `libtock` `examples` and `tests` folders, so that we should each mostly have examples that match the drivers. I moved a few drivers up or down when the alphabetic approach wasn't perfect. I think this should work well enough.

# @hudson-ayers 
Drivers
```
 libtock/adc.c                                   |  81 ++---
 libtock/adc.h                                   |   2 +-
 libtock/alarm_timer.c                           |  32 +-
 libtock/ambient_light.c                         |  18 +-
 libtock/analog_comparator.c                     |  45 +--
 libtock/analog_comparator.h                     |   6 +-
 libtock/app_state.c                             |  14 +-
 libtock/ble.c                                   |  44 +--
 libtock/button.c                                |  32 +-
 libtock/button.h                                |  10 +-
 libtock/buzzer.c                                |  32 +-
 libtock/console.c                               |  56 ++--
 libtock/crc.c                                   |  50 ++-
 libtock/dac.c                                   |  10 +-
 libtock/dac.h                                   |   4 +-
 libtock/gpio.c                                  |  71 +----
 libtock/gpio.h                                  |   4 +-
 libtock/gpio_async.c                            |  94 +-----
 libtock/humidity.c                              |  16 +-
 libtock/i2c_master.c                            |  80 ++---
 libtock/i2c_master_slave.c                      |  95 ++----
 libtock/i2c_master_slave.h                      |   6 +-
 libtock/ieee802154.c                            | 388 ++++++++++--------------
 libtock/ieee802154.h                            | 388 ++++++++++--------------
```

Examples:
```
 examples/adc/main.c                                |  12 +-
 examples/ble_advertising/main.c                    |  10 +-
 examples/ble_passive_scanning/main.cpp             |   6 +-
 examples/blink/main.c                              |   4 +-
 examples/buttons/main.c                            |  10 +-
 .../courses/2018-11-SenSys/sensys_udp_rx/main.c    |   8 +-
 examples/find_north/main.c                         |  16 +-
 examples/ip_sense/main.c                           |   5 +-
 examples/l3gd20/main.c                             |  38 +--
```

Tests:
```
 examples/tests/adc/main.c                          |   6 +-
 examples/tests/adc_continuous/main.c               |  22 +-
 examples/tests/adc_single_samples/main.c           |   4 +-
 examples/tests/ambient_light/main.c                |   2 +-
 examples/tests/analog_comparator/main.c            |   9 +-
 examples/tests/ble/ble_advertise/main.c            |   2 +-
 examples/tests/ble/ble_gap_library/main.c          |   8 +-
 examples/tests/ble/ble_test_tx_power/main.c        |   4 +-
 examples/tests/button_print/main.c                 |   3 +-
 examples/tests/callback_remove_test01/main.c       |   6 +-
 examples/tests/console/main.c                      |   2 +-
 examples/tests/console_recv_long/main.c            |   2 +-
 examples/tests/console_recv_short/main.c           |   2 +-
 examples/tests/console_timeout/main.c              |   2 +-
 examples/tests/crc/main.c                          |   8 +-
 examples/tests/dac/main.c                          |   6 +-
 examples/tests/gpio/main.c                         |   3 +-
 examples/tests/gpio_loopback_test/main.c           |  21 +-
 examples/tests/hail/main.c                         |  15 +-
 examples/tests/i2c/i2c_master_ping_pong/main.c     |   9 +-
 .../tests/i2c/i2c_master_slave_ping_pong/main.c    |  32 +--
 examples/tests/ieee802154/radio_ack/main.c         |   4 +-
 examples/tests/ieee802154/radio_tx/main.c          |   4 +-
```

# @lschuermann :
```
 libtock/internal/alarm.h                        |   7 ++-
 libtock/internal/alarm_internal.c               |  31 ++++-------
 libtock/internal/nonvolatile_storage.h          |  10 ++--
 libtock/internal/nonvolatile_storage_internal.c |  52 ++++++------------
 libtock/ipc.c                                   |  52 +++++++-----------
 libtock/ipc.h                                   |   6 +-
 libtock/l3gd20.c                                | 141 +++++++++++++++++++++++------------------------
 libtock/led.c                                   |  26 ++-------
 libtock/led.h                                   |   2 +-
 libtock/lps25hb.c                               |  21 ++-----
 libtock/lps25hb.h                               |   2 +-
 libtock/lsm303dlhc.c                            | 192 +++++++++++++++++++++++++++++++++-------------------------------
 libtock/ltc294x.c                               | 138 +++++++++++-----------------------------------
 libtock/ltc294x.h                               |   6 +-
 libtock/max17205.c                              | 103 +++++-----------------------------
 libtock/ninedof.c                               |  40 ++++----------
 libtock/nrf51_serialization.c                   |  45 ++++-----------
 libtock/pca9544a.c                              |  58 +++++---------------
```

Examples:
```
 examples/lsm303dlhc/main.c                         |  68 +++---
 examples/lvgl/lvgl_driver.c                        |   4 +-
 examples/lvgl/main.c                               |   2 +-
 examples/music/main.c                              |  46 ++--
 examples/rot13_client/main.c                       |   6 +-
 examples/screen/main.c                             |  81 ++++---
 examples/sensors/main.c                            |  28 +--
```

Tests:
```
 examples/tests/imix/main.c                         |   3 +-
 examples/tests/lps25hb/main.c                      |   4 +-
 examples/tests/max17205/main.c                     |   8 +-
 examples/tests/mpu_walk_region/main.c              |   7 +-
 examples/tests/multi_alarm_test/main.c             |   3 +-
 examples/tests/nonvolatile_storage/main.c          |  27 ++-
 examples/tests/number_guess_game/main.c            |   7 +-
```


# @alevy 
```
 libtock/proximity.h                             |  30 +-
 libtock/proximity.c                             |  30 +-
 libtock/rng.c                                   |  63 ++--
 libtock/rng.h                                   |  11 +-
 libtock/screen.c                                | 356 +++++++++++++---------
 libtock/screen.h                                |  12 +-
 libtock/sdcard.c                                |  71 ++---
 libtock/sound_pressure.c                        |  34 +--
 libtock/spi.c                                   | 143 ++++-----
 libtock/spi_peripheral.c                        | 104 +++----
 libtock/sys.c                                   |   2 +-
 libtock/temperature.c                           |  19 +-
 libtock/text_screen.c                           | 206 +++++++------
 libtock/tock.c                                  | 133 +++++---
 libtock/tock.h                                  | 162 ++++++----
 libtock/touch.c                                 |  58 ++--
 libtock/touch.h                                 |   2 +-
 libtock/tsl2561.c                               |  23 +-
 libtock/tsl2561.h                               |   2 +-
 libtock/udp.c                                   | 102 +++----
 libtock/udp.h                                   |   2 +-
 libtock/unit_test.c                             |   5 +-
 libtock/usb.c                                   |  38 +--
```

Examples:
```
 .../ble-env-sense/test-with-sensors/main.c         |   4 +-
 examples/services/ble-env-sense/test/main.c        |   4 +-
 .../ble/ble_nrf5x_concurrency/Advertiser2/main.c   |   4 +-
 .../ble/ble_nrf5x_concurrency/Advertiser3/main.c   |   4 +-
 .../ble/ble_nrf5x_concurrency/Advertiser4/main.c   |   4 +-
 examples/text_screen/main.c                        |   4 +-
 examples/touch/main.c                              |   8 +-
 examples/tutorials/02_button_print/main.c          |   3 +-
 examples/tutorials/05_ipc/led/main.c               |   2 +-
 examples/tutorials/05_ipc/logic/main.c             |  10 +-
 examples/tutorials/05_ipc/rng/main.c               |   3 +-
```

Tests:
```
 examples/tests/proximity/main.c                    |   3 +-
 examples/tests/rng/main.c                          |   3 +-
 examples/tests/spi/spi_controller_transfer/main.c  |   3 +-
 examples/tests/tsl2561/main.c                      |   5 +-
 examples/tests/udp/udp_rx/main.c                   |   8 +-
 examples/tests/udp/udp_send/main.c                 |   2 +-
 examples/tests/udp/udp_virt_app_kernel/main.c      |   2 +-
 examples/tests/udp/udp_virt_app_tests/app1/main.c  |   4 +-
 examples/tests/udp/udp_virt_app_tests/app2/main.c  |   4 +-
 examples/tests/udp/udp_virt_rx_tests/app1/main.c   |   4 +-
 examples/tests/udp/udp_virt_rx_tests/app2/main.c   |   4 +-
 examples/tests/usb/main.c                          |   2 +-
 examples/tests/yield_for_with_timeout_test/main.c  |   4 +-
 examples/unit_tests/nrf52840dk-test/main.c         | 268 +++++++++++----------
```

File: libtock/max17205.c [hunk 6]
```diff
@@ -12,25 +12,23 @@ static struct max17205_data result   = { .fired = false, .rc = 0, .value0 = 0, .
...
 
   struct max17205_data* data = (struct max17205_data*) ud;
-  data->rc     = return_code;
+  data->rc     = tock_status_to_returncode(status);
   data->value0 = value0;
   data->value1 = value1;
   data->fired  = true;
 }
 
-static bool is_busy = false;
```
Comment (lschuermann) [inline libtock/max17205.c:27]: These changes (removing the check of whether a pending upcall has been issued by the capsules) seem unrelated to the essence of this PR. Is this done just because this isn't typically done in libtock-c driver API implementations? I'm fine with leaving it in here, but I'm curious to know whether there's an actual requirement because of the type / wrapper changes of this PR to remove these checks.

**Review by lschuermann — COMMENTED** [review]
I closely inspected the files designated for me to review. In general, I think these changes are pretty awesome and the wrapper functions improve the code a lot, as before the system call error handling was pretty different from one driver to another. I think some changes don't directly relate to the essence of this PR, but are mostly small formatting or code style improvements.

I did not test these changes yet, in part because I don't have many of the required hardware components and given it was decided to postpone full testing of these large changes to release testing. I will try to execute some tests nonetheless.

Also, I really _really_ wished C would have a more strict type system. :laughing:

File: libtock/ieee802154.c [hunk 7]
```diff
@@ -67,196 +67,138 @@ bool ieee802154_is_up(void) {
...
-  } else if (com.type < TOCK_SYSCALL_SUCCESS) {
-    return tock_error_to_rcode(com.data[0]);
-  } else {
-    // incorrect success type
-    return TOCK_FAIL;
-  }
+  int ret = tock_command_return_u32_to_returncode(com, (uint32_t*) addr);
+
+  // Driver adds 1 to make the value positive.
+  *addr -= 1;
```
Comment (hudson-ayers) [inline libtock/ieee802154.c:111]: outside of the scope of this PR, but would be nice to remove tricks like this necessitated by ReturnCode in the kernel

File: libtock/ieee802154.c [hunk 8]
```diff
@@ -268,66 +210,55 @@ int ieee802154_get_neighbor(unsigned index,
...
-    return com.data[0];
-  } else if (com.type < TOCK_SYSCALL_SUCCESS) {
-    return tock_error_to_rcode(com.data[0]);
-  } else {
-    return TOCK_FAIL;
-  }
+  int ret = tock_command_return_u32_to_returncode(com, (uint32_t*) level);
+
+  // Driver adds 1 to ensure it is positive.
+  *level -= 1;
```
Comment (hudson-ayers) [inline libtock/ieee802154.c:259]: In the case of this command returning an error, this decrements an uninitialized variable. The caller shouldn't use the returned value anyway, but I think this is technically UB. I don't think it would matter in practice though.

File: examples/tests/ble/ble_test_tx_power/main.c [hunk 9]
```diff
@@ -25,13 +25,13 @@ int main(void) {
...
       err = ble_set_tx_power(i);
-      if (err < TOCK_SUCCESS) {
+      if (err < RETURNCODE_SUCCESS) {
         printf("Test failed \t power_level %d could not be configured\r\n", i);
         return 0;
       }
     } else {
       err = ble_set_tx_power(i);
-      if (err >= TOCK_SUCCESS) {
+      if (err >= RETURNCODE_SUCCESS) {
```
Comment (hudson-ayers) [inline examples/tests/ble/ble_test_tx_power/main.c:34]:
```suggestion
      if (err == RETURNCODE_SUCCESS) {
```

It should no longer be possible for this function to return a value greater than 0 (nit)

**Review by hudson-ayers — COMMENTED** [review]
I have gone through all of my files now. Once these comments are addressed I am happy with this PR.

Overall, this is a great PR. The Tock 2.0 transition left a lot of drivers in pretty inconsistent states, and this gets things to a much better state.

File: libtock/max17205.c (hunk 6 above)
Comment (bradjc) [reply libtock/max17205.c:27]: Nope. Just bad practice to store a copy of `is_busy` here (and won't work if there is more than one app anyway). Since I'm (likely) never going to go through every file in libtock-c again, decided to fix now rather than leave the bug.

File: libtock/alarm_timer.c [hunk 10]
```diff
@@ -123,21 +124,14 @@ void alarm_cancel(alarm_t* alarm) {
...
 
-uint32_t alarm_read(void) {
-  syscall_return_t rval = command(DRIVER_NUM_ALARM, 2, 0, 0);
-  if (rval.type == TOCK_SYSCALL_SUCCESS_U32) {
-    return rval.data[0];
-  } else {
-    return 0;
-  }
-}
-
```
[the remark this answers is not in the document]
Comment (bradjc) [reply libtock/alarm_timer.c:134]: Because I could I guess? This is a direct syscall, and those are handled by alarm_internal.

File: libtock/ieee802154.c (hunk 7 above)
Comment (bradjc) [reply libtock/ieee802154.c:111]: I have no idea what this is doing.

File: examples/tests/ble/ble_test_tx_power/main.c (hunk 9 above)
Comment (bradjc) [reply examples/tests/ble/ble_test_tx_power/main.c]: I think I fixed this correctly.

File: libtock/nrf51_serialization.c [hunk 11]
```diff
@@ -8,63 +8,42 @@
...
 
 
 
 int nrf51_serialization_write(char* tx, int tx_len) {
   // Pass in the TX buffer.
   allow_ro_return_t aval = allow_readonly(DRIVER_NUM_NRF_SERIALIZATION, 0, tx, tx_len);
-  if (!aval.success) {
-    return tock_error_to_rcode(aval.error);
-  }
+  if (!aval.success) return tock_status_to_returncode(aval.status);
```
[the remark this answers is not in the document]
Comment (bradjc) [reply libtock/nrf51_serialization.c:33]: I think these middle-of-function checks ended up like this in many places.

**Comment by bradjc** [comment]
Addressed comments.

File: examples/tests/ble/ble_test_tx_power/main.c (hunk 9 above)
Comment (hudson-ayers) [reply examples/tests/ble/ble_test_tx_power/main.c]: I don't think so. This test is actually verifying that if you try to set an invalid power level an error is returned

File: examples/tests/ble/ble_test_tx_power/main.c [hunk 12]
```diff
@@ -25,13 +25,13 @@ int main(void) {
...
       err = ble_set_tx_power(i);
-      if (err < TOCK_SUCCESS) {
+      if (err < RETURNCODE_SUCCESS) {
         printf("Test failed \t power_level %d could not be configured\r\n", i);
         return 0;
       }
     } else {
       err = ble_set_tx_power(i);
-      if (err >= TOCK_SUCCESS) {
+      if (err < RETURNCODE_SUCCESS) {
```
Comment (bradjc) [inline examples/tests/ble/ble_test_tx_power/main.c:34]:
```suggestion
      if (err == RETURNCODE_SUCCESS) {
```

File: examples/tests/ble/ble_test_tx_power/main.c (hunk 9 above)
Comment (bradjc) [reply examples/tests/ble/ble_test_tx_power/main.c]: Ok, gotcha, I didn't understand the printf comment. Fixed now.

File: libtock/ieee802154.c (hunk 7 above)
Comment (hudson-ayers) [reply libtock/ieee802154.c:111]: yeah, lets leave this for later. But I think it was because we were using SuccessWithValue, but sometimes the value we wanted to pass was 0, which would be indistinguishable from Success (or something like that)

File: libtock/alarm_timer.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (hudson-ayers) [reply libtock/alarm_timer.c:134]: sounds good

File: libtock/max17205.c (hunk 6 above)
Comment (lschuermann) [reply libtock/max17205.c:27]: Thanks for fixing this!

**Review by alevy — APPROVED** [review]

**Comment by alevy** [comment]
I just had one minor nit, and I don't actually care if it's adopted.

**Review by hudson-ayers — APPROVED** [review]
